<a href="https://colab.research.google.com/github/hamiin47/OB-session/blob/main/Assign_0404.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **07. 문서 군집화 소개와 실습 ( Opinion Review 데이터 세트 )**

**문서 군집화 개념**
- 문서 군집화는 비슷한 텍스트 구성의 문서를 군집화하는 것이다.
- 동일한 군집에 속하는 문서를 같은 카테고리 소속으로 분류할 수 있으므로 앞에서 소개한 텍스트 분류 기반의 문서 분류와 유사하다.
- 그러나 텍스트 분류 기반의 문서 분류는 사전에 결정 카테고리 값을 가진 학습 데이터 세트가 필요한 데 반해, 문서 군집화는 학습 데이터 세트가 필요 없는 비지도 학습 기반으로 동작한다.


---



**Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기**


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd
import glob, os
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

# 디렉토리 설정
path = r'/content/drive/MyDrive/ESAA/OB/csv/OpinosisDataset1.0/topics'

# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, '*.data'))
filename_list = []
opinion_text = []

# 개별 파일들의 파일명을 filename_list 리스트로 취합
# 개별 파일들의 파일 내용은 DataFrame 로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합
for file_ in all_files:
  # 개별 파일을 읽어서 DataFrame으로 생성
  df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
  # 절대경로로 주어진 파일명을 가공. Linux에서 수행 시에는 아래 /를 \로 변경
  # 맨 마지막 .data 확장자도 제거
  filename_ = file_.split('/')[-1]
  filename = filename_.split('.')[0]

  # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가
  filename_list.append(filename)
  opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를 DataFrame으로 생성
document_df = pd.DataFrame({'filename': filename_list, 'opinion_text': opinion_text})
document_df.head()

filename  \
0   display_garmin_nuvi_255W_gps   
1     battery-life_ipod_nano_8gb   
2     battery-life_amazon_kindle   
3  accuracy_garmin_nuvi_255W_gps   
4      comfort_honda_accord_2008   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...  
1                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...  
2                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...  
3                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...  
4                                                                                                                                                                                                                                                                                         

In [4]:
# 커스텀 어근 변환 함수 LemNormalize() 함수 생성
import nltk
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')

def LemNormalize(text):
  wordnet_lemmatizer = WordNetLemmatizer()
  return [wordnet_lemmatizer.lemmatize(token) for token in text.split()]

[nltk_data] Downloading package wordnet to /root/nltk_data...


- 문서를 TF-IDF 형태로 피처 벡터화 진행
- TfidVectorizer는 Lemmatization 같은 어근 변환을 직접 지원하진 않지만, tokenizer 인자에 커스텀 어근 변환 함수를 적용해 어근 변환을 수행할 수 있다.

In [5]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english',
                             ngram_range=(1,2), min_df=0.05, max_df=0.85)

# opinion_text 칼럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

In [6]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터일 결과 도출용 random_state=0
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

In [7]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0   display_garmin_nuvi_255W_gps   
1     battery-life_ipod_nano_8gb   
2     battery-life_amazon_kindle   
3  accuracy_garmin_nuvi_255W_gps   
4      comfort_honda_accord_2008   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...   
1                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
2                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
3                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
4                                                                                                                                                                                                                                                                                    

In [8]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
12  interior_honda_accord_2008   
27  interior_toyota_camry_2007   
19   quality_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
12                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
27                                                                                                                                                                                                                                      First of all, the interior has way too many cheap plastic parts like the cheap plastic center piece that houses the clock .\n0                                                                                                                                                                       3 blown struts at 30,000 miles, interior trim coming loose and rattling squeaking, stains on paint, and bug splats taking paint off, premature uneven brake wear, on 3rd windsh...   
19                                                                                                                                                                                                                                                                        I previously owned a Toyota 4Runner which had incredible build quality and reliability .\n0                                                                                                                                                                                                                                                                                               I bought the Camry because of Toyota reliability and qua...   

    cluster_label  
12              0  
27              0  
19              0

In [9]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
9    bathroom_bestwestern_hotel_sfo   
26          food_holiday_inn_london   
21           food_swissotel_chicago   
22       free_bestwestern_hotel_sfo   
17   location_bestwestern_hotel_sfo   
30      location_holiday_inn_london   
18    parking_bestwestern_hotel_sfo   
11         price_holiday_inn_london   
16          room_holiday_inn_london   
20      rooms_bestwestern_hotel_sfo   
31          rooms_swissotel_chicago   
36    service_bestwestern_hotel_sfo   
43       service_holiday_inn_london   
39  service_swissotel_hotel_chicago   
41      staff_bestwestern_hotel_sfo   
48          staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
9                                                                                                                                                                                                                                                 The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
26                                                                                                                                                                                                                                                                      The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
21                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
22                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still

In [10]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
3     accuracy_garmin_nuvi_255W_gps   
2        battery-life_amazon_kindle   
1        battery-life_ipod_nano_8gb   
6       battery-life_netbook_1005ha   
5             buttons_amazon_kindle   
7   directions_garmin_nuvi_255W_gps   
0      display_garmin_nuvi_255W_gps   
28    eyesight-issues_amazon_kindle   
13                features_windows7   
15              fonts_amazon_kindle   
25          keyboard_netbook_1005ha   
24         navigation_amazon_kindle   
32              price_amazon_kindle   
40   satellite_garmin_nuvi_255W_gps   
42      screen_garmin_nuvi_255W_gps   
50             screen_ipod_nano_8gb   
37            screen_netbook_1005ha   
34         size_asus_netbook_1005ha   
47              sound_ipod_nano_8gb   
45       speed_garmin_nuvi_255W_gps   
49                   speed_windows7   
33     updates_garmin_nuvi_255W_gps   
35              video_ipod_nano_8gb   
38       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
3                                                                                                                                                                                                                                                                  , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
2                                                                                                                                                                                                                After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
1                                                                                                                                                                                                                                                                                                  short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
6                                                                                                                                                                                                                              

In [11]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  \
4   comfort_honda_accord_2008   
8   comfort_toyota_camry_2007   
46    seats_honda_accord_2008   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
4                                                                                                                                                                                                                                                                                                                                                                                                 Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
8                                                                                                                                                   Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
46                                                                                                                                                                                                                                                                                      Front seats are very uncomfortable .\n0                                                                                                                                                                                                                                  No memory seats, no trip computer, can only display outside temp with trip odometer .\n1                                                                   ...   

    cluster_label  
4               3  
8               3  
46              3

In [12]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  \
14   gas_mileage_toyota_camry_2007   
23       mileage_honda_accord_2008   
29   performance_honda_accord_2008   
10      performance_netbook_1005ha   
44  transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
14                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
23                                                                                                                                                                                It's quiet, get good gas mileage and looks clean inside and out .\n0                                                                                                                                                                     The mileage is great, and I've had to get used to stopping less for gas .\n1                                                                                                                                                                                                         Thought gas ...   
29                                                                                     Very happy with my 08 Accord, performance is quite adequate it has nice looks and is a great long,  distance cruiser .\n0                                                                                                                                 6, 4, 3 eco engine has poor performance and gas mileage of 22 highway .\n1                                                                                                                                                 Overall performance is good but comfort level is poor .\n2                                                                                      ...   
10                                                                                                            The Eee Super Hybrid Engine utility lets users overclock or underclock their Eee PC's to boost performance or provide better battery life depending on their immediate requirements .\n0                                                                                                                                                                                                                  In Super Performance mode CPU, Z shows the bus speed to increase up to 169 .\n1                                                                                                                  One...   
44                                                                                                                                                                                                                                                                    

- 전반적으로 군집 개수가 약간 많게 설정되어 있음
- 중심 개수를 5개에서 3개로 낮춰 진행

In [13]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# 소속 클러스터를 cluster_label 칼럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
4         comfort_honda_accord_2008   
14    gas_mileage_toyota_camry_2007   
12       interior_honda_accord_2008   
8         comfort_toyota_camry_2007   
27       interior_toyota_camry_2007   
29    performance_honda_accord_2008   
23        mileage_honda_accord_2008   
19        quality_toyota_camry_2007   
46          seats_honda_accord_2008   
44   transmission_toyota_camry_2007   
16          room_holiday_inn_london   
17   location_bestwestern_hotel_sfo   
9    bathroom_bestwestern_hotel_sfo   
11         price_holiday_inn_london   
20      rooms_bestwestern_hotel_sfo   
18    parking_bestwestern_hotel_sfo   
31          rooms_swissotel_chicago   
26          food_holiday_inn_london   
30      location_holiday_inn_london   
43       service_holiday_inn_london   
36    service_bestwestern_hotel_sfo   
39  service_swissotel_hotel_chicago   
21           food_swissotel_chicago   
22       free_bestwestern_hotel_sfo   
41      staff_bestwestern_hotel_sfo   
48          staff_swissotel_chicago   
0      display_garmin_nuvi_255W_gps   
13                features_windows7   
15              fonts_amazon_kindle   
5             buttons_amazon_kindle   
6       battery-life_netbook_1005ha   
7   directions_garmin_nuvi_255W_gps   
28    eyesight-issues_amazon_kindle   
24         navigation_amazon_kindle   
25          keyboard_netbook_1005ha   
3     accuracy_garmin_nuvi_255W_gps   
1        battery-life_ipod_nano_8gb   
10       performance_netbook_1005ha   
2        battery-life_amazon_kindle   
32              price_amazon_kindle   
38       voice_garmin_nuvi_255W_gps   
37            screen_netbook_1005ha   
34         size_asus_netbook_1005ha   
35              video_ipod_nano_8gb   
42      screen_garmin_nuvi_255W_gps   
40   satellite_garmin_nuvi_255W_gps   
45       speed_garmin_nuvi_255W_gps   
33     updates_garmin_nuvi_255W_gps   
47              sound_ipod_nano_8gb   
49                   speed_windows7   
50             screen_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
4                                                                                                                                                                                                                                                                                                                                                                                                 Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
14                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                              

**군집별 핵심 단어 추출하기**

In [14]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape:', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape: (3, 5648)
[[0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.00043948 0.00099222 ... 0.00197    0.00156757 0.00156757]
 [0.00171135 0.00126592 0.         ... 0.         0.         0.        ]]


군집 3개, word 피처 5648개로 구성

- 각 행의 배열 값은 각 군집 내의 피처의 위치가 개별 중심과 얼마나 가까운가를 상대 값으로 나타냄
- 0과 1까지의 값을 가질 수 있으며 1에 가까울수록 중심과 가까움

In [15]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명을 반환
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}

    # cluster.centers array의 값이 큰 순으로 정렬된 인덱스 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:, ::-1]

    # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster']=cluster_num

        # cluster_centers_.argsort()[:,::-1]로 구한 인덱스를 이용해 top n 피처 단어를 구함
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [feature_names[ind] for ind in top_feature_indexes]

        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()

        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심 위치 상댓값, 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()

        cluster_details[cluster_num]['filenames'] = filenames

    return cluster_details

In [16]:
#더 보기 좋게 표현하기 위한 별도의 print_cluster_details() 함수
def print_cluster_details(cluster_details):
  for cluster_num, cluster_detail in cluster_details.items():
    print('##### Cluster {0}'.format(cluster_num))
    print('Top features:', cluster_detail['top_features'])
    print('Reviews 파일명:', cluster_detail['filenames'][:7])
    print('==============================================')

In [17]:
# 피처명 리스트는 tfidf_vect 객체에서 get_feature_names_out()로 추출
feature_names = tfidf_vect.get_feature_names_out()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,
                                      feature_names=feature_names, clusters_num=3, top_n_features=10)
print_cluster_details(cluster_details)

##### Cluster 0
Top features: ['interior', 'seat', 'mileage', 'gas', 'gas mileage', 'comfortable', 'transmission', 'car', 'quality', 'performance']
Reviews 파일명: ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'interior_honda_accord_2008', 'gas_mileage_toyota_camry_2007', 'quality_toyota_camry_2007', 'mileage_honda_accord_2008', 'interior_toyota_camry_2007']
##### Cluster 1
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'price', 'parking', 'clean']
Reviews 파일명: ['bathroom_bestwestern_hotel_sfo', 'price_holiday_inn_london', 'room_holiday_inn_london', 'location_bestwestern_hotel_sfo', 'parking_bestwestern_hotel_sfo', 'rooms_bestwestern_hotel_sfo', 'food_swissotel_chicago']
##### Cluster 2
Top features: ['screen', 'battery', 'keyboard', 'kindle', 'video', 'battery life', 'voice', 'direction', 'size', 'life']
Reviews 파일명: ['display_garmin_nuvi_255W_gps', 'battery-life_ipod_nano_8gb', 'battery-life_amazon_kindle', 'accuracy_garmin_nuvi_255W_gps